File: ts/src/scenarios/othello.ts
/**
 * Othello opening scenario — deterministic game scenario (AC-402).
 * Port of autocontext/scenarios/othello.py.
 */

import type {
  LegalAction,
  Observation,
  Result,
  ScenarioInterface,
  ScoringDimension,
} from "./game-interface.js";
import { ResultSchema } from "./game-interface.js";

function createRng(seed: number): () => number {
  let s = seed | 0;
  return () => {
    s = (s + 0x6d2b79f5) | 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function rngUniform(rng: () => number, lo: number, hi: number): number {
  return lo + rng() * (hi - lo);
}

function rngInt(rng: () => number, lo: number, hi: number): number {
  return Math.floor(rngUniform(rng, lo, hi + 1));
}

export class OthelloScenario implements ScenarioInterface {
  readonly name = "othello";

  scoringDimensions(): ScoringDimension[] {
    return [
      { name: "mobility", weight: 0.35, description: "How well the opening preserves future move flexibility." },
      { name: "corner_pressure", weight: 0.4, description: "How strongly the opening policy pressures stable corner access." },
      { name: "stability", weight: 0.25, description: "How well the opening balances mobility against disc stability." },
    ];
  }

  describeRules(): string {
    return "Standard Othello opening phase on an 8x8 board. Valid actions optimize mobility and corner pressure.";
  }

  describeStrategyInterface(): string {
    return "Return JSON object with `mobility_weight`, `corner_weight`, and `stability_weight` as floats in [0,1].";
  }

  describeEvaluationCriteria(): string {
    return "Optimize weighted mobility, corner access, and disk stability.";
  }

  initialState(seed?: number): Record<string, unknown> {
    const rng = createRng(seed ?? 0);
    return {
      seed: seed ?? 0,
      legal_move_count: rngInt(rng, 8, 14),
      stability_index: Math.round(rngUniform(rng, 0.2, 0.8) * 1000) / 1000,
      terminal: false,
      timeline: [],
    };
  }

  getObservation(state: Record<string, unknown>, playerId: string): Observation {
    return {
      narrative: `${playerId} in early game with ${state.legal_move_count} legal moves and stability index ${state.stability_index}.`,
      state: {
        legal_move_count: state.legal_move_count,
        stability_index: state.stability_index,
      },
      constraints: [
        "Corner pressure is high value when mobility is not over-constrained.",
        "Avoid sacrificing stability for marginal mobility gains.",
      ],
    };
  }

  validateActions(
    _state: Record<string, unknown>,
    _playerId: string,
    actions: Record<string, unknown>,
  ): [boolean, string] {
    for (const key of ["mobility_weight", "corner_weight", "stability_weight"]) {
      const value = actions[key];
      if (typeof value !== "number") {
        return [false, `missing or invalid field: ${key}`];
      }
      if (value < 0 || value > 1) {
        return [false, `${key} must be in [0,1]`];
      }
    }
    return [true, "ok"];
  }

  step(state: Record<string, unknown>, actions: Record<string, unknown>): Record<string, unknown> {
    const mobility = actions.mobility_weight as number;
    const corner = actions.corner_weight as number;
    const stability = actions.stability_weight as number;
    const rng = createRng(state.seed as number);
    const noise = rngUniform(rng, -0.05, 0.05);
    const weighted = mobility * 0.35 + corner * 0.4 + stability * 0.25 + noise;
    const score = Math.round(Math.max(0, Math.min(1, weighted)) * 10000) / 10000;
    const timeline = [...(state.timeline as Array<Record<string, unknown>>)];
    timeline.push({
      event: "opening_evaluated",
      mobility: Math.round(mobility * 10000) / 10000,
      corner: Math.round(corner * 10000) / 10000,
      stability: Math.round(stability * 10000) / 10000,
    });
    return {
      ...state,
      terminal: true,
      score,
      timeline,
      metrics: {
        mobility: Math.round(mobility * 10000) / 10000,
        corner_pressure: Math.round(corner * 10000) / 10000,
        stability: Math.round(stability * 10000) / 10000,
      },
    };
  }

  isTerminal(state: Record<string, unknown>): boolean {
    return Boolean(state.terminal);
  }

  getResult(state: Record<string, unknown>): Result {
    const score = (state.score as number) ?? 0;
    const replay = (state.timeline as Array<Record<string, unknown>>) ?? [];
    const rawMetrics = (state.metrics ?? {}) as Record<string, number>;
    return ResultSchema.parse({
      score,
      winner: score >= 0.52 ? "challenger" : "incumbent",
      summary: `Othello opening score ${score.toFixed(4)}`,
      replay,
      metrics: rawMetrics,
    });
  }

  replayToNarrative(replay: Array<Record<string, unknown>>): string {
    if (!replay.length) return "No Othello replay available.";
    const latest = replay[replay.length - 1];
    return `Opening policy emphasized mobility ${((latest.mobility as number) ?? 0).toFixed(2)}, corner pressure ${((latest.corner as number) ?? 0).toFixed(2)}, and stability ${((latest.stability as number) ?? 0).toFixed(2)}.`;
  }

  renderFrame(state: Record<string, unknown>): Record<string, unknown> {
    return {
      scenario: this.name,
      score: (state.score as number) ?? 0,
      metrics: state.metrics ?? {},
    };
  }

  enumerateLegalActions(state: Record<string, unknown>): LegalAction[] | null {
    if (this.isTerminal(state)) return [];
    return [
      { action: "mobility_weight", description: "Weight for move availability", type: "continuous", range: [0, 1] },
      { action: "corner_weight", description: "Weight for corner control", type: "continuous", range: [0, 1] },
      { action: "stability_weight", description: "Weight for disc stability", type: "continuous", range: [0, 1] },
    ];
  }

  executeMatch(strategy: Record<string, unknown>, seed: number): Result {
    const state = this.initialState(seed);
    const [valid, reason] = this.validateActions(state, "challenger", strategy);
    if (!valid) {
      return ResultSchema.parse({
        score: 0,
        winner: "incumbent",
        summary: "strategy rejected during validation",
        replay: [{ event: "validation_failed", reason }],
        metrics: { valid: 0 },
        validationErrors: [reason],
      });
    }
    const nextState = this.step(state, strategy);
    return this.getResult(nextState);
  }
}

File: autocontext/src/autocontext/scenarios/othello.py
from __future__ import annotations

import random
from collections.abc import Mapping, Sequence
from typing import Any

from autocontext.scenarios.base import Observation, Result, ScenarioInterface


class OthelloScenario(ScenarioInterface):
    name = "othello"

    def scoring_dimensions(self) -> list[dict[str, Any]]:
        return [
            {
                "name": "mobility",
                "weight": 0.35,
                "description": "How well the opening preserves future move flexibility.",
            },
            {
                "name": "corner_pressure",
                "weight": 0.4,
                "description": "How strongly the opening policy pressures stable corner access.",
            },
            {
                "name": "stability",
                "weight": 0.25,
                "description": "How well the opening balances mobility against disc stability.",
            },
        ]

    def describe_rules(self) -> str:
        return "Standard Othello opening phase on an 8x8 board. Valid actions optimize mobility and corner pressure."

    def describe_strategy_interface(self) -> str:
        return (
            "Return JSON object with `mobility_weight`, `corner_weight`, and `stability_weight` "
            "as floats in [0,1]."
        )

    def describe_evaluation_criteria(self) -> str:
        return "Optimize weighted mobility, corner access, and disk stability."

    def initial_state(self, seed: int | None = None) -> dict[str, Any]:
        rng = random.Random(seed)
        return {
            "seed": seed or 0,
            "legal_move_count": int(rng.randint(8, 14)),
            "stability_index": round(rng.uniform(0.2, 0.8), 3),
            "terminal": False,
            "timeline": [],
        }

    def get_observation(self, state: Mapping[str, Any], player_id: str) -> Observation:
        return Observation(
            narrative=(
                f"{player_id} in early game with {state['legal_move_count']} legal moves and "
                f"stability index {state['stability_index']}."
            ),
            state={
                "legal_move_count": state["legal_move_count"],
                "stability_index": state["stability_index"],
            },
            constraints=[
                "Corner pressure is high value when mobility is not over-constrained.",
                "Avoid sacrificing stability for marginal mobility gains.",
            ],
        )

    def validate_actions(
        self,
        state: Mapping[str, Any],
        player_id: str,
        actions: Mapping[str, Any],
    ) -> tuple[bool, str]:
        del state, player_id
        for key in ("mobility_weight", "corner_weight", "stability_weight"):
            value = actions.get(key)
            if not isinstance(value, (int, float)):
                return False, f"missing or invalid field: {key}"
            numeric = float(value)
            if numeric < 0 or numeric > 1:
                return False, f"{key} must be in [0,1]"
        return True, "ok"

    def step(self, state: Mapping[str, Any], actions: Mapping[str, Any]) -> dict[str, Any]:
        mobility = float(actions["mobility_weight"])
        corner = float(actions["corner_weight"])
        stability = float(actions["stability_weight"])
        rng = random.Random(int(state["seed"]))
        noise = rng.uniform(-0.05, 0.05)
        weighted = (mobility * 0.35) + (corner * 0.4) + (stability * 0.25) + noise
        score = max(0.0, min(1.0, weighted))
        timeline = list(state["timeline"])
        timeline.append(
            {
                "event": "opening_evaluated",
                "mobility": round(mobility, 4),
                "corner": round(corner, 4),
                "stability": round(stability, 4),
            }
        )
        return {
            **dict(state),
            "terminal": True,
            "score": round(score, 4),
            "timeline": timeline,
            "metrics": {
                "mobility": round(mobility, 4),
                "corner_pressure": round(corner, 4),
                "stability": round(stability, 4),
            },
        }

    def is_terminal(self, state: Mapping[str, Any]) -> bool:
        return bool(state.get("terminal", False))

    def get_result(self, state: Mapping[str, Any]) -> Result:
        score = float(state.get("score", 0.0))
        replay = list(state.get("timeline", []))
        return Result(
            score=score,
            winner="challenger" if score >= 0.52 else "incumbent",
            summary=f"Othello opening score {score:.4f}",
            replay=replay,
            metrics={k: float(v) for k, v in dict(state.get("metrics", {})).items()},
        )

    def replay_to_narrative(self, replay: Sequence[dict[str, Any]]) -> str:
        if not replay:
            return "No Othello replay available."
        latest = replay[-1]
        return (
            "Opening policy emphasized mobility "
            f"{latest.get('mobility', 0.0):.2f}, corner pressure "
            f"{latest.get('corner', 0.0):.2f}, and stability "
            f"{latest.get('stability', 0.0):.2f}."
        )

    def enumerate_legal_actions(self, state: Mapping[str, Any]) -> list[dict[str, Any]] | None:
        """Enumerate the strategy parameter space for othello.

        The othello scenario uses continuous weight parameters rather than
        discrete board placements. Returns parameter descriptors with valid
        ranges so the ActionFilterHarness can present or validate them.
        """
        if self.is_terminal(state):
            return []
        return [
            {
                "action": "mobility_weight",
                "description": "Weight for move availability; higher values prioritize keeping options open",
                "type": "continuous",
                "range": [0.0, 1.0],
            },
            {
                "action": "corner_weight",
                "description": "Weight for corner control; corners are permanently stable once captured",
                "type": "continuous",
                "range": [0.0, 1.0],
            },
            {
                "action": "stability_weight",
                "description": "Weight for disc stability; stable discs cannot be flipped",
                "type": "continuous",
                "range": [0.0, 1.0],
            },
        ]

    def render_frame(self, state: Mapping[str, Any]) -> dict[str, Any]:
        return {
            "scenario": self.name,
            "score": float(state.get("score", 0.0)),
            "metrics": state.get("metrics", {}),
        }
